core/src/main/scala/caliban/execution/Executor.scala

```diff
@@ -306,27 +306,43 @@ object Executor {
     ZIO.succeed(GraphQLResponse(NullValue, List(error)))
 
   private[caliban] def mergeFields(field: Field, typeName: String): List[Field] = {
-    val map      = new java.util.LinkedHashMap[String, Field]()
-    var modified = false
-
-    field.fields.foreach { field =>
-      if (field._condition.forall(_.contains(typeName))) {
-        map.compute(
-          field.aliasedName,
-          (_, f) =>
-            if (f == null) field
-            else {
-              modified = true
-              f.copy(fields = f.fields ::: field.fields)
-            }
-        )
-      } else {
-        modified = true
+    def haveSameCondition(head: Field, tail: List[Field]): Boolean = {
+      val condition = head._condition
+      var remaining = tail
+      while (!remaining.isEmpty) {
+        if (remaining.head._condition != condition) return false
```

Comment:
I was gonna suggest tail recursion but this will probably iterate tighter

Reply:
Hmm I think the compiler will rewrite the tailrec method likely to this implementation. I'll check, if it's the same I'll use tailrec

Reply:
Looking at the generated code, it looks like `@tailrec` uses Java labels in an infinite with return statements, whereas this one uses the method as a condition to the `while` loop. Not sure which one would be more performant in theory

Reply:
Yeah we're probably getting into a grey area of compiler optimization. I think the conditionless while might be able to skip a branch if equal instruction, though for modern chips branches aren't really performance killers anymore